[Issue 243] Move to new metadata spec

dalogsdon opened the pull request:
Part of #243

parente commented:
Fixes #253 and #235 as a side effects.

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata.js:
```diff
@@ -17,160 +17,149 @@ define([
+    }
+
+    // get cell view metadata from element data
+    function _getCellViewMetadata($elem, viewId) {
```
Maybe `_getCellActiveViewMetadata`?

dalogsdon replied:
It's not just for the active view. The third argument specifies the view to get and defaults to the active view. See line 145 (_updateCellMetadata calls with 3rd arg) for an example of 3rd arg being used.

parente replied:
I see it now. Maybe add some doc on it then indicating the default behavior of the param (2nd right?) is missing.

parente commented:
Tested that none of the recent bugs have resurfaced: #222, #223, #226, #227, #231, #233, #234, #239, #248. 👍

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata.js:
```diff
@@ -17,160 +17,149 @@ define([
-            layout = metadata.layout;
+        // common metadata
+        _ensureHas(metadata, 'extensions');
+        _ensureHas(metadata.extensions, 'jupyter_dashboards');
```
Why not update `_ensureHas` to take the full dot-delimited string, similar to what [dojo.getObject](https://dojotoolkit.org/reference-guide/1.7/dojo/getObject.html) does?  i.e. `_ensureHas(metadata, 'extensions.jupyter_dashboards')`.

dalogsdon commented:
Updated based on comments and to add compatibility module.

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata-compatibility.js:
```diff
@@ -0,0 +1,100 @@
+         * This must be run AFTER the version 1 metadata is initialized in
+         * `dashboard-metadata.js`.
+         */
+        convert: function() {
```
If we're going to play the up-convert game, I think we should future proof a bit. Should we make this more amenable to v0 -> v1 conversion explicitly so that later, another convert function / delegate can hand v1 -> v2 and so on?

parente replied:
Nevermind for now. I see there's an if-statement guarding the convert so that it only runs on v0 (urth) metadata. If a metadata v2 ever happens, older versions of the extension won't try to run their v0->v1 code, and v1->v2 conversion is something we can add to a new version of the extension. For now, YAGNI.

parente commented:
@dalogsdon I can't tell because of the squash push, but in your latest commit, did you do anything other than address @jhpedemonte 's comment and add the compatibility support? Need to know how much I have to go back and retest. (Push separate commits if more changes are coming ...)

dalogsdon commented:
Updated with @jhpedemonte's comment about ensuring object exists, updated @parente's comment on the `_getCellViewMetadata` function documentation, and the compatibility module.

Because the ensure object exist util is used in the generation of the metadata structure, some re-testing may be good to ensure no bugs with the structure.

parente commented:
I'll give it one more round of testing.

parente commented:
Tested all of the recent defects again as well as loading of many existing demo notebooks. No issues found.
